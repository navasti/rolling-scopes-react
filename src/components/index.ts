export { Loader } from './Loader';
export { Modal } from './Modal';
