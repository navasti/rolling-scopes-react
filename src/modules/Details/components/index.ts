export { MoveCardDetails } from './MoveCardDetails';
export { TypeCardDetails } from './TypeCardDetails';
export { PokemonCardDetails } from './PokemonCardDetails';
