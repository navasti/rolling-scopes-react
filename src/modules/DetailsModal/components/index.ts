export { TypeModalContent } from './TypeModalContent';
export { MoveModalContent } from './MoveModalContent';
export { PokemonModalContent } from './PokemonModalContent';
