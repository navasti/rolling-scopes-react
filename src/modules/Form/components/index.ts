export { BirthdayField } from './BirthdateField';
export { ErrorMessage } from './ErrorMessage';
export { ConsentField } from './ConsentField';
export { AvatarField } from './AvatarField';
export { GenderField } from './GenderField';
export { ShinyField } from './ShinyField';
export { TypeField } from './TypeField';
export { NameField } from './NameField';
export { FormCard } from './FormCard';
