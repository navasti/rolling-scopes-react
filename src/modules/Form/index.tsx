import React, { createRef, FormEvent, RefObject } from 'react';
import { CheckFields, CustomPokemon, ErrorsObject, TextFields } from 'types';
import { Layout } from 'modules';
import * as S from './styled';
import { uuid } from 'utils';
import {
  BirthdayField,
  ErrorMessage,
  ConsentField,
  GenderField,
  AvatarField,
  ShinyField,
  TypeField,
  NameField,
  FormCard,
} from 'modules/Form/components';
import { FIELDS_VALIDATION_BY_NAME, FEMALE, MALE, ErrorMessages } from 'appConstants';

type Props = {
  componentName: string;
  location: string;
};

type State = {
  pokemons: Array<CustomPokemon & { id: string }>;
  initialEnter: boolean;
  errors: ErrorsObject;
};

export class Form extends React.Component<Props, State> {
  femaleInputRef: RefObject<HTMLInputElement>;
  successMessage: RefObject<HTMLSpanElement>;
  birthdayField: RefObject<HTMLInputElement>;
  isShinyField: RefObject<HTMLInputElement>;
  consentField: RefObject<HTMLInputElement>;
  maleInputRef: RefObject<HTMLInputElement>;
  avatarField: RefObject<HTMLInputElement>;
  typeField: RefObject<HTMLSelectElement>;
  nameField: RefObject<HTMLInputElement>;

  constructor(props: Props) {
    super(props);
    this.state = {
      pokemons: [],
      initialEnter: true,
      errors: {
        birthday: null,
        consent: null,
        gender: null,
        name: null,
        type: null,
      },
    };
    this.successMessage = createRef();
    this.femaleInputRef = createRef();
    this.birthdayField = createRef();
    this.consentField = createRef();
    this.isShinyField = createRef();
    this.maleInputRef = createRef();
    this.avatarField = createRef();
    this.typeField = createRef();
    this.nameField = createRef();
  }

  removeError = (field: TextFields | CheckFields) => {
    if (this.state.initialEnter) this.setState({ initialEnter: false });
    const errors = { ...this.state.errors };
    errors[field] = null;
    this.setState({ errors });
  };

  validate = (fields: Array<React.RefObject<HTMLInputElement | HTMLSelectElement>>) => {
    const errors = { ...this.state.errors };
    const textFields = FIELDS_VALIDATION_BY_NAME.TEXT.map((textField) =>
      fields.find((field) => field.current?.name === textField)
    );
    const checkFields = FIELDS_VALIDATION_BY_NAME.CHECK.map((checkField) =>
      fields.filter((field) => field.current?.name === checkField)
    );
    checkFields.forEach((checkField) => {
      const valid = checkField.some((field) => (field.current as HTMLInputElement)?.checked);
      const name = checkField[0]?.current?.name
        ? (checkField[0].current.name as CheckFields)
        : null;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    textFields.forEach((textField) => {
      const name = textField?.current?.name ? (textField.current.name as TextFields) : null;
      const value = textField?.current?.value;
      const valid = value && value.trim().length >= 2;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    return errors;
  };

  prepareCustomPokemon = () => {
    const shiny = !!this.isShinyField.current?.checked;
    const female = this.femaleInputRef?.current?.checked;
    const file = this.avatarField.current?.files?.[0] || null;
    const customPokemon: CustomPokemon & { id: string } = {
      birthday: this.birthdayField?.current?.value || '',
      type: this.typeField?.current?.value || '',
      name: this.nameField?.current?.value || '',
      gender: female ? FEMALE : MALE,
      avatar: file,
      id: uuid(),
      shiny,
    };
    return customPokemon;
  };

  handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const fields = [
      this.nameField,
      this.typeField,
      this.avatarField,
      this.isShinyField,
      this.consentField,
      this.birthdayField,
      this.maleInputRef,
      this.femaleInputRef,
    ];
    const errors = this.validate(fields);
    if (Object.values(errors).some((error) => error != null)) {
      this.setState({ errors });
    } else {
      const customPokemon = this.prepareCustomPokemon();
      this.setState({
        pokemons: [...this.state.pokemons, customPokemon],
      });
      fields.forEach((field) => {
        const current = field.current;
        if (current) {
          if (current.type === 'radio' || current.type === 'checkbox') {
            (current as HTMLInputElement).checked = false;
          } else current.value = '';
        }
      });
    }
  };

  render() {
    const { location, componentName } = this.props;
    return (
      <Layout location={location} componentName={componentName}>
        <S.CommonView>
          <S.FormHeading>Create custom pokemon!</S.FormHeading>
          <S.Form onSubmit={this.handleSubmit}>
            <NameField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.nameField}
            />
            <TypeField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.typeField}
            />
            <S.GenderWrapper>
              <S.RadioWrapper>
                *Gender
                <S.RadioFields>
                  <GenderField
                    name={FEMALE}
                    ref={this.femaleInputRef}
                    onChange={this.removeError}
                  />
                  <GenderField name={MALE} ref={this.maleInputRef} onChange={this.removeError} />
                </S.RadioFields>
              </S.RadioWrapper>
              <ErrorMessage
                visible={!!this.state.errors.gender}
                message={this.state.errors.gender}
              />
            </S.GenderWrapper>
            <BirthdayField
              ref={this.birthdayField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <AvatarField ref={this.avatarField} />
            <ShinyField ref={this.isShinyField} />
            <ConsentField
              ref={this.consentField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <S.SubmitButton
              disabled={
                this.state.initialEnter ||
                Object.values(this.state.errors).some((error) => error != null)
              }
              type="submit"
            >
              Submit
            </S.SubmitButton>
            <S.SuccessMessage ref={this.successMessage}>Pokemon added successfuly</S.SuccessMessage>
          </S.Form>
          <hr />
          <S.CardsGrid>
            {this.state.pokemons.map((pokemon) => (
              <FormCard key={pokemon.id} customPokemon={pokemon} />
            ))}
          </S.CardsGrid>
        </S.CommonView>
      </Layout>
    );
  }
}
