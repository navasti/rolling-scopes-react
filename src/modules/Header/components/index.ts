export { LocationBadge } from './LocationBadge';
export { Navigation } from './Navigation';
