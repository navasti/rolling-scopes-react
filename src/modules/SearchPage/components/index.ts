export { SearchBar } from './SearchBar';
export { Cards } from './Cards';
export { Card } from './Card';
export { Tabs } from './Tabs';
