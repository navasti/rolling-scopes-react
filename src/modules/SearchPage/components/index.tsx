export { SearchBar } from './SearchBar';
export { Card } from './Card';
