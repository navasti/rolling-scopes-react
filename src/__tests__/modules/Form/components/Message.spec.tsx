import { screen, render } from '@testing-library/react';
import { Message } from 'modules/Form/components';
import { MessageType } from 'types';

describe('Message', () => {
  it('provided message should be displayed and match snapshot', () => {
    const { container } = render(
      <Message message="testing-message" type={MessageType.success} visible center={false} />
    );
    expect(screen.getByText('Testing-message')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
  it('color of message should be red for error type message', () => {
    render(<Message message="testing-message" type={MessageType.error} visible center={false} />);
    expect(screen.getByText('Testing-message')).toHaveStyle({ color: 'red' });
  });
  it('color of message should be green for success type message', () => {
    render(<Message message="testing-message" type={MessageType.success} visible center={false} />);
    expect(screen.getByText('Testing-message')).toHaveStyle({ color: 'green' });
  });
  it('message should not be visible if visible prop is equal to false', () => {
    render(
      <Message message="testing-message" type={MessageType.error} visible={false} center={false} />
    );
    expect(screen.getByText('Testing-message')).toHaveStyle({ opacity: 0 });
  });
  it('message should be centered if center prop is equal to true', () => {
    render(<Message message="testing-message" type={MessageType.error} visible center={true} />);
    expect(screen.getByText('Testing-message')).toHaveStyle({ textAlign: 'center' });
  });
});
