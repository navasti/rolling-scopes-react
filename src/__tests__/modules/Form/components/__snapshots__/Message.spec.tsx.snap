// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message provided message should be displayed and match snapshot 1`] = `
<div>
  <span
    class="sc-kDDrLX egFLGi"
    type="success"
  >
    Testing-message
  </span>
</div>
`;
